src/upgrades/mod.rs

```diff
@@ -86,30 +146,101 @@ impl From<crate::commons::error::Error> for UpgradeError {
+}
+
+impl Default for DataUpgradeInfo {
+    fn default() -> Self {
+        Self {
+            last_event: 0,
+            last_command: 0,
+            last_update: Time::now(), // will be overwritten to appropriate value
+        }
+    }
+}
+
+impl From<&DataUpgradeInfo> for StoredValueInfo {
+    fn from(upgrade_info: &DataUpgradeInfo) -> Self {
+        StoredValueInfo {
+            snapshot_version: 0,
+            last_event: upgrade_info.last_event,
+            last_command: upgrade_info.last_command,
+            last_update: upgrade_info.last_update,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum UpgradeMode {
+    PrepareOnly,
+    PrepareThenFinalise,
+}
+
 //------------ UpgradeStore --------------------------------------------------
 
 /// Implement this for automatic upgrades to key stores
 pub trait UpgradeStore {
     fn needs_migrate(&self) -> Result<bool, UpgradeError>;
-    fn migrate(&self) -> Result<(), UpgradeError>;
 
-    fn version_before(kv: &KeyValueStore, later: KrillVersion) -> Result<bool, UpgradeError> {
-        kv.version_is_before(later).map_err(UpgradeError::KeyStoreError)
+    fn prepare_new_data(&self, mode: UpgradeMode) -> Result<(), UpgradeError>;
+
+    fn version_before(&self, later: KrillVersion) -> Result<bool, UpgradeError> {
+        self.store()
+            .version_is_before(later)
+            .map_err(UpgradeError::KeyStoreError)
     }
 
     fn store(&self) -> &KeyValueStore;
 
+    fn new_store(&self) -> &KeyValueStore;
+
+    fn data_upgrade_info_key(scope: &str) -> KeyStoreKey {
+        KeyStoreKey::scoped(scope.to_string(), "upgrade_info.json".to_string())
+    }
+
+    /// Return the DataUpgradeInfo telling us to where we got to with this migration.
+    fn data_upgrade_info(&self, scope: &str) -> UpgradeResult<DataUpgradeInfo> {
+        self.new_store()
+            .get(&Self::data_upgrade_info_key(scope))
+            .map(|opt| match opt {
+                None => DataUpgradeInfo::default(),
+                Some(info) => info,
+            })
+            .map_err(UpgradeError::KeyStoreError)
+    }
+
+    /// Update the DataUpgradeInfo
+    fn update_data_upgrade_info(&self, scope: &str, info: &DataUpgradeInfo) -> UpgradeResult<()> {
+        self.new_store()
+            .store(&Self::data_upgrade_info_key(scope), info)
+            .map_err(UpgradeError::KeyStoreError)
+    }
+
+    /// Removed the DataUpgradeInfo
+    fn remove_data_upgrade_info(&self, scope: &str) -> UpgradeResult<()> {
+        self.new_store()
+            .drop_key(&Self::data_upgrade_info_key(scope))
+            .map_err(UpgradeError::KeyStoreError)
+    }
+
     // Find all command keys and sort them by sequence.
     // Then turn them back into key store keys for further processing.
-    fn command_keys(&self, scope: &str) -> Result<Vec<KeyStoreKey>, UpgradeError> {
+    fn command_keys(&self, scope: &str, from: u64) -> Result<Vec<KeyStoreKey>, UpgradeError> {
```

Comment:
Is the comment about "find _all_ command keys" still correct if you now filter based on the given "from" ?